### PR TITLE
Build fail on error

### PR DESCRIPTION
**Pull request by sladyn98**

### CODE REVIEW

File: src/main/java/io/jenkins/plugins/analysis/core/steps/IssuesPublisher.java [hunk 1]
```diff
@@ -48,12 +48,14 @@
...
     private final LogHandler logger;
+    private final IssuesRecorder issuesRecorder;
 
     @SuppressWarnings("ParameterNumber")
     IssuesPublisher(final Run<?, ?> run, final AnnotatedReport report,
             final HealthDescriptor healthDescriptor, final QualityGateEvaluator qualityGate,
             final String name, final String referenceJobName, final boolean ignoreQualityGate,
-            final boolean ignoreFailedBuilds, final Charset sourceCodeEncoding, final LogHandler logger) {
+            final boolean ignoreFailedBuilds, final Charset sourceCodeEncoding, final LogHandler logger
+    ,final IssuesRecorder issuesRecorder) {
```
Comment (uhafner) [inline src/main/java/io/jenkins/plugins/analysis/core/steps/IssuesPublisher.java]: Rather than using parameter `issueRecorder` you need to provide the boolean `failOnErrors `.

File: src/main/java/io/jenkins/plugins/analysis/core/steps/IssuesPublisher.java [hunk 2]
```diff
@@ -85,6 +88,10 @@ ResultAction attachAction() {
         logger.log("Created analysis result for %d issues (found %d new issues, fixed %d issues)",
                 result.getTotalSize(), result.getNewSize(), result.getFixedSize());
 
+        if(issuesRecorder.getFailOnErrors() && report.getReport().hasErrors()) {
+            // ADD CODE TO FAIL BUILD
+        }
+
```
Comment (uhafner) [inline src/main/java/io/jenkins/plugins/analysis/core/steps/IssuesPublisher.java]:
```
if (failOnErrors && report.getReport().hasErrors()) {
   QualityGateStatusHandler...
}
```

File: src/main/java/io/jenkins/plugins/analysis/core/steps/IssuesPublisher.java (hunk 1 above)
Comment (sladyn98) [reply src/main/java/io/jenkins/plugins/analysis/core/steps/IssuesPublisher.java]: But is that value not set in `IssuesRecorder` and can be obtained from its getter ?

File: src/main/java/io/jenkins/plugins/analysis/core/steps/IssuesPublisher.java (hunk 2 above)
Comment (sladyn98) [reply src/main/java/io/jenkins/plugins/analysis/core/steps/IssuesPublisher.java]: Done +1

**Comment by sladyn98** [comment]
@uhafner  kindly review

File: src/main/java/io/jenkins/plugins/analysis/core/steps/IssuesRecorder.java [hunk 3]
```diff
@@ -460,14 +461,15 @@ public Descriptor getDescriptor() {
         return (Descriptor) super.getDescriptor();
     }
 
-    @Override
-    public boolean perform(final AbstractBuild<?, ?> build, final Launcher launcher, final BuildListener listener)
+    
+    public boolean perform(final AbstractBuild<?, ?> build, final Launcher launcher, final BuildListener listener,
```
[the remark this answers is not in the document]
Comment (uhafner) [reply src/main/java/io/jenkins/plugins/analysis/core/steps/IssuesRecorder.java]: `failOnError` has to be a field as well here (with data bound setters). This property will be set in the UI. (This part is still missing, see [config.jelly](https://github.com/jenkinsci/warnings-ng-plugin/blob/master/src/main/resources/io/jenkins/plugins/analysis/core/steps/IssuesRecorder/config.jelly)). I thing you had that already before your new changes.

**Comment by uhafner** [comment]
If the `IssueRecorder` is running as expected, you need also make the same change to the 2 steps that use the `IssuePublisher` method.

File: src/main/java/io/jenkins/plugins/analysis/core/steps/IssuesRecorder.java [hunk 4]
```diff
@@ -540,7 +563,7 @@ private String getReportName(final Tool tool) {
     }
 
     private AnnotatedReport scanWithTool(final Run<?, ?> run, final FilePath workspace, final TaskListener listener,
-            final Tool tool) throws IOException, InterruptedException {
+                                         final Tool tool) throws IOException, InterruptedException {
```
Comment (uhafner) [inline src/main/java/io/jenkins/plugins/analysis/core/steps/IssuesRecorder.java]: is this formatting correct? Why did you change this line, it seems unrelated?

File: src/main/java/io/jenkins/plugins/analysis/core/steps/IssuesRecorder.java [hunk 5]
```diff
@@ -578,7 +601,7 @@ private Charset getCharset(final String encoding) {
      */
     @SuppressWarnings("deprecation")
     void publishResult(final Run<?, ?> run, final TaskListener listener, final String loggerName,
-            final AnnotatedReport report, final String reportName, final StageResultHandler statusHandler) {
+                       final AnnotatedReport report, final String reportName, final StageResultHandler statusHandler) {
```
Comment (uhafner) [inline src/main/java/io/jenkins/plugins/analysis/core/steps/IssuesRecorder.java]: is this formatting correct? Why did you change this line, it seems unrelated?

File: src/main/java/io/jenkins/plugins/analysis/core/steps/IssuesRecorder.java [hunk 6]
```diff
@@ -494,7 +517,7 @@ private String createLoggerPrefix() {
     }
 
     private void record(final Run<?, ?> run, final FilePath workspace, final TaskListener listener,
-            final StageResultHandler statusHandler)
+                        final StageResultHandler statusHandler)
```
Comment (uhafner) [inline src/main/java/io/jenkins/plugins/analysis/core/steps/IssuesRecorder.java]: is this formatting correct? Why did you change this line, it seems unrelated?

File: src/main/java/io/jenkins/plugins/analysis/core/steps/IssuesRecorder.java [hunk 7]
```diff
@@ -477,7 +500,7 @@ public boolean perform(final AbstractBuild<?, ?> build, final Launcher launcher,
      * that has Pipeline-specific behavior.
      */
     void perform(@NonNull final Run<?, ?> run, @NonNull final FilePath workspace,
-            @NonNull final TaskListener listener, @NonNull final StageResultHandler statusHandler)
+                 @NonNull final TaskListener listener, @NonNull final StageResultHandler statusHandler)
```
Comment (uhafner) [inline src/main/java/io/jenkins/plugins/analysis/core/steps/IssuesRecorder.java]: is this formatting correct? Why did you change this line, it seems unrelated?

File: src/main/resources/io/jenkins/plugins/analysis/core/steps/IssuesRecorder/config.jelly [hunk 8]
```diff
@@ -17,6 +17,11 @@
     <f:checkbox />
   </f:entry>
 
+   <f:entry title="${%title.failBuildOnError}" description="${%description.failBuildOnError}" field="failBuildOnError">
+      <f:checkbox />
+    </f:entry>
+
+
```
Comment (uhafner) [inline src/main/resources/io/jenkins/plugins/analysis/core/steps/IssuesRecorder/config.jelly]: Sorry, this is my fault: we should place this entry into the file [publish-parameters.jelly](https://github.com/jenkinsci/warnings-ng-plugin/blob/master/src/main/resources/issues/publish-parameters.jelly) (at line 7). Then it will be used automatically by the `IssuesRecorder`, `PublishIssuesStep`, `RecordIssuesStep`.

File: src/main/resources/io/jenkins/plugins/analysis/core/steps/IssuesRecorder/config.properties [hunk 9]
```diff
@@ -5,4 +5,7 @@ title.enabledForFailure=Run always
 description.enabledForFailure=Enable recording for failed builds.
 
 title.aggregatingResults=Aggregate results
-description.aggregatingResults=Aggregate results of all tools into a single result.
\ No newline at end of file
+description.aggregatingResults=Aggregate results of all tools into a single result.
```
Comment (uhafner) [inline src/main/resources/io/jenkins/plugins/analysis/core/steps/IssuesRecorder/config.properties]: The same here, into `publish-parameters.properties`

File: src/main/resources/io/jenkins/plugins/analysis/core/steps/IssuesRecorder/help-buildOnError.html [hunk 10]
```diff
@@ -0,0 +1,2 @@
+By default, static analysis results are only recorded for stable or unstable builds, but not for failed builds:
```
Comment (uhafner) [inline src/main/resources/io/jenkins/plugins/analysis/core/steps/IssuesRecorder/help-buildOnError.html]: I'm not sure if this works if it is placed at the publish-parameters level. Maybe we need it 3 times, for `IssuesRecorder`, `PublishIssuesStep`, `RecordIssuesStep`.

File: src/main/java/io/jenkins/plugins/analysis/core/steps/IssuesRecorder.java (hunk 4 above)
Comment (sladyn98) [reply src/main/java/io/jenkins/plugins/analysis/core/steps/IssuesRecorder.java]: Yeah I added an argument there which was unnecessary.Removed it now.

File: src/main/java/io/jenkins/plugins/analysis/core/steps/IssuesRecorder.java (hunk 6 above)
Comment (sladyn98) [reply src/main/java/io/jenkins/plugins/analysis/core/steps/IssuesRecorder.java]: Removed :+1:

File: src/main/java/io/jenkins/plugins/analysis/core/steps/IssuesRecorder.java (hunk 7 above)
Comment (sladyn98) [reply src/main/java/io/jenkins/plugins/analysis/core/steps/IssuesRecorder.java]: Removed :+1:

File: src/main/resources/io/jenkins/plugins/analysis/core/steps/IssuesRecorder/help-buildOnError.html (hunk 10 above)
Comment (sladyn98) [reply src/main/resources/io/jenkins/plugins/analysis/core/steps/IssuesRecorder/help-buildOnError.html]: Yeah , lets add it three times and remove if not required :)

File: src/main/resources/io/jenkins/plugins/analysis/core/steps/IssuesRecorder/config.jelly (hunk 8 above)
Comment (sladyn98) [reply src/main/resources/io/jenkins/plugins/analysis/core/steps/IssuesRecorder/config.jelly]: Resolved :+1:

File: src/main/resources/io/jenkins/plugins/analysis/core/steps/IssuesRecorder/config.properties (hunk 9 above)
Comment (sladyn98) [reply src/main/resources/io/jenkins/plugins/analysis/core/steps/IssuesRecorder/config.properties]: Resolved :+1:

**Comment by sladyn98** [comment]
@uhafner Kindly review

File: src/main/resources/issues/publish-parameters.properties [hunk 11]
```diff
@@ -6,6 +6,8 @@ title.ignoreFailedBuilds=Ignore Failed Builds
 description.ignoreFailedBuilds=If checked, then only successful or unstable reference builds \
   will be considered. This option is enabled by default, since analysis results might be inaccurate if the build failed. \
   If unchecked, every build that contains a static analysis result is considered, even if the build failed.
+title.failOnError=Fail Build on Error.
```
Comment (uhafner) [inline src/main/resources/issues/publish-parameters.properties]: `Fail on errors` (and no point)

File: src/main/resources/issues/publish-parameters.properties [hunk 12]
```diff
@@ -6,6 +6,8 @@ title.ignoreFailedBuilds=Ignore Failed Builds
...
+description.failOnError=Fails Build if there is an error during job execution.
```
Comment (uhafner) [inline src/main/resources/issues/publish-parameters.properties]: Fails the build if errors have been reported during the execution of this step.

File: src/test/java/io/jenkins/plugins/analysis/warnings/recorder/MiscIssuesRecorderITest.java [hunk 13]
```diff
@@ -235,6 +233,25 @@ public void shouldHaveOriginsIfBuildContainsWarnings() {
...
     }
 
+    /**
+     * This test is used to check if the build fails when the setFailOnError has been set to true.
+     */
+    @Test
+    public void shouldFailBuildWhenFailBuildOnErrorsIsSet() {
+        FreeStyleProject job = createFreeStyleProject();
+        enableEclipseWarnings(job);
+
```
Comment (uhafner) [inline src/test/java/io/jenkins/plugins/analysis/warnings/recorder/MiscIssuesRecorderITest.java]: Let the job run here and verify that it is successful.
You can also check the `AnalysisResuls` for error messages so we see that this is the actual cause for the failure in line 252.

File: src/test/java/io/jenkins/plugins/analysis/warnings/recorder/MiscIssuesRecorderITest.java [hunk 14]
```diff
@@ -235,6 +233,25 @@ public void shouldHaveOriginsIfBuildContainsWarnings() {
         assertThat(result).hasQualityGateStatus(QualityGateStatus.INACTIVE);
     }
 
+    /**
+     * This test is used to check if the build fails when the setFailOnError has been set to true.
+     */
```
Comment (uhafner) [inline src/test/java/io/jenkins/plugins/analysis/warnings/recorder/MiscIssuesRecorderITest.java]: Add a sentence why there are errors in the build or why the build fails.

File: src/main/resources/io/jenkins/plugins/analysis/core/steps/RecordIssuesStep/help-buildOnError.html [hunk 15]
```diff
@@ -0,0 +1,2 @@
+By default, static analysis results are only recorded for stable or unstable builds, but not for failed builds:
+analysis results might be inaccurate if the build failed. If you want the build to fail on error kindly tick this checkbox.
```
Comment (uhafner) [inline src/main/resources/io/jenkins/plugins/analysis/core/steps/RecordIssuesStep/help-buildOnError.html]: If there are errors while scanning the log or files for issues (e.g., file pattern matches no files, source files could not be copied, etc.) then the warning plugin will show these errors in a separate view but does not alter the build state. If you would rather like to fail the build on such errors, please tick this checkbox.

**Comment by uhafner** [comment]
Thanks a lot!

**Comment by sladyn98** [comment]
Your Welcome :)
